vowpalwabbit/parse_example.cc

```diff
@@ -4,17 +4,45 @@
 
 #include <cmath>
 #include <cctype>
+#include <future>
+#include <thread>
+#include <chrono>
 #include "parse_example.h"
 #include "hash.h"
 #include "unique_sort.h"
 #include "global_data.h"
 #include "constant.h"
 #include "vw_string_view.h"
+#include "io_to_queue.h"
+
+class IO_Item;
 
 size_t read_features(vw* all, char*& line, size_t& num_chars)
 {
-  line = nullptr;
-  size_t num_chars_initial = readto(*(all->p->input), line, '\n');
+ 
+  IO_Item result;
+
+  while(true){
+
+    if((*all).p->_io_state.done_with_io && (*all).p->_io_state.io_lines->size() == 0){
+      break;
+    }
+
+    result = pop_io_queue(all);
+
+    if(result.numCharsInit > 0){
+      break;
+    }
+
+  }
+
+  std::string result_string =  result.message;
+
+  line = new char[result_string.size() + 1];
```

Comment:
This is not ideal since we'd have to deal with the memory cleanup in multiple places (text parser + json parser + maybe cache parser). Ideally we'd convert line to a string instead of a char*, but the json parser requires a mutable c-string as its input.

I'm checking what our options are, at worst we may be able to make a small (but really awful) change as a **_temporary_** band-aid

Reply:
I spoke with the guy who wrote the json parser and he was working on a change that makes this a lot nicer. He's trying to get it into a place where its usable, once thats done I'll pass it over to you so we can start building on top of it.

Essentially we'll be changing `char*& line` to `std::string& line` where the caller of this function will pass in a reference to the string to have it filled. Keep this the way it is for now since we'll have to refactor it anyways once the new interface is determined